### DISCUSSION TITLE
Allow using operator in a developer setting

### DISCUSSION THREAD

**Discussion by ctron** [keycloak/keycloak#11944] [Archive]
Assuming I am building some application, on top of the Keycloak operator, I might want to offer my users also a "developer" deployment. Something that doesn't require HTTPS and can run on e.g. Minikube.

However, I also want to re-use the Keycloak operator "as is", and not be required to manually install Keycloak in my developer deployment, just because the Keycloak operator only supports a production grade deployment.

Maybe that can be achieved but some global `.spec.developerMode: true` flag? Or by some other explicit configuration, acknowledging by the user that this might not be secure.

In any case, I wouldn't prevent the user from doing this. As you might never know what the intent of the user was.

**Comment by andreaTP** [comment]
Thanks a lot for opening this discussion!
I think that, what you are describing, is a legit use-case that we, eventually, want to support ( also @lburgazzoli stumbled on something similar).

Regarding the implementation we have a few possible alternatives:

 1. an environment variable in the operator: `KEYCLOAK_OPERATOR_DEV_MODE`
    PRO: this makes it possible to re-use CRs in-between `dev` and `prod` (possibly with no changes)
    CONS: not immediately visible, hard to explain
    
  2. a CR field `.spec.unsupported.devMode`
     PRO: it's the most "obvious" implementation, easy to document and implement
     CONS: interaction with the rest of the (especially `required`) top-level fields is going to be complicated
     
   3. an alternative CRD `KeycloakDev`
     PRO: makes the use-case a first-class citizen and the distinction is clear, it will also work well for cluster-wide installations
     CONS: more maintenance/complexity for something that is not the focus of the project, the user still need to use different resources

**Reply by andreaTP** [reply]
cc. @vmuzikar @pedroigor @DGuhr

**Reply by ctron** [reply]
I think that option 2 is the best. Although I wouldn't name it "unsupported". If you consider additional options for dev mode only, maybe use:

```yaml
spec:
  devMode:
    enabled: true
```

And for more options:

```yaml
spec:
  devMode:
    enabled: true
    someExtraOption: "someExtraValue"
```

That way one needs to explicitly enable it. It follows the idea of CRDs, is extensible, and reasonably easy to implement. And you can always claim that there is no support for this section of the CRD. Maybe even set a warning on the status section of the `Keycloak` resource if this is enabled.

**Comment by vmuzikar** [comment]
@ctron Hi and thank you for raising this topic and your interest in the new Operator!

I believe it really depends on the use-case of the Operator. As you correctly stated, the Operator currently targets only production-grade deployments. This is and always will be the main use-case. It is a good question whether the Operator should support other use-cases as well, as e.g. the stated dev mode.

Could you please elaborate a little on what is your use-case for the dev mode? Why do you want to use the Operator in this scenario?

Our current expectations for now have been that for dev purposes and playing around with the Keycloak container, users can use e.g. plain Docker. For development of the Operator as such, we usually just use our [example K8s objects](https://github.com/keycloak/keycloak/tree/main/operator/src/main/resources).

**Reply by ctron** [reply]
> Could you please elaborate a little on what is your use-case for the dev mode? Why do you want to use the Operator in this scenario?

As I tried to explain above. We do not use plain docker for developers. As we targeting Kubernetes, that doesn't seem really feasible. We are incorporating components like KNative, Strimzi, … and Keycloak. For a developer experience and CI, we rely on tools like Kind, Minikube, and CRC. All of them targeting developers.

If we would not be able to deploy Keycloak in a developer mode, that would mean we would need to either re-create the operator for our development mode (in which case all the testing would not be done with the actual Keycloak operator, but be performed with ours). Or we would need to use an alternative to Keycloak for developers and CI.

In both cases, one might ask: why do you use Keycloak anyway?

So again, I think it is reasonable to have a "production ready" default. But don't block people in doing something different.

**Reply by andreaTP** [reply]
> So again, I think it is reasonable to have a "production ready" default. But don't block people in doing something different.

I agree with this sentence, but, as you experienced, we are not "actively blocking" users doing anything differently.

At the same time, I do believe that yours is a valid use case, and there are quite a few low-hanging fruits we can tackle to improve without compromising the main goal of "production-ready deployments".

**Reply by vmuzikar** [reply]
@ctron Thank you for clarifying! Makes total sense.

I'd like to discuss a bit the specific steps that we could take here. You are mentioning HTTPS and Minikube. We actually [do support disabling the HTTPS](https://www.keycloak.org/operator/advanced-configuration#_disabling_required_cr_fields). And the Operator of course can be deployed on Minishift, that's what we're actually using for its development. :) Are there any other blockers for you for using the Operator in a dev environment?

**Reply by ctron** [reply]
This originates back on the comment from here: https://github.com/keycloak/keycloak/issues/11926#issuecomment-1123521664

Indeed, technically it might not need much. However, it should be documented in a way: need this? do that! … Maybe in the context of a "developer deployment".

**Reply by vmuzikar** [reply]
Maybe we could just explicitly document it in the [README](https://github.com/keycloak/keycloak/blob/main/operator/README.md). WDYT @ctron @andreaTP?

**Reply by andreaTP** [reply]
My honest feeling is that the setup for "development" is still too verbose to be "just documented", but happy to hear other opinions!

**Reply by vmuzikar** [reply]
@andreaTP My thoughts were that since we're talking about development purposes (primarily of the Operator itself but possibly in more general, as stated by @ctron), and AFAICT we already have all the tools (disabling SSL/hostname), we could just create some hints for easier development in the README. Similarly to e.g. [what the old Operator has](https://github.com/keycloak/keycloak-operator#local-development).

**Reply by andreaTP** [reply]
@vmuzikar that would be one step for sure, I think that we can do a little more by enabling an "easier" path for developers "out-of-the-box" with one of the proposals made [here](https://github.com/keycloak/keycloak/discussions/11944#discussioncomment-2731588).

**Reply by ADiTuri** [reply]
Is there a similar option to [disable https](https://github.dev/keycloak/keycloak-operator/blob/main/deploy/crds/keycloak.org_keycloaks_crd.yaml) also in v1alpha1 operator? Struggle to find it

**Reply by andreaTP** [reply]
@ADiTuri this discussion is regarding the new operator.
With the old one you have "insecure" material for HTTPs by default, but please, open an issue there if you have any issue with it.
